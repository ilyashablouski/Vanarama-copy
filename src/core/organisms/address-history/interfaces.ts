import { IBaseProps } from '../../interfaces/base';
import { IDropdown } from '../../interfaces/dropdown';
import { IAddress } from '../../interfaces/address';

interface IPropertyStatuses extends IDropdown {}

export interface IAddressHistoryObject {
  address: IAddress;
  propertyStatus: string;
  month: number;
  year: number;
}

export interface IAddressHistoryProps extends IBaseProps {
  name: string;
  minYearsOfHistory: number;
  lookBackYears: number;
  propertyStatuses: IPropertyStatuses;
  addressHistory?: IAddressHistoryObject[];
  setValue?: Function; // Callback.
}

export interface IAddressHistoryState {
  addressHistory?: IAddressHistoryObject[];
  completed: boolean;
  allValid: boolean;
}
