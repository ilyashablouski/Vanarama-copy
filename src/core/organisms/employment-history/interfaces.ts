import { IBaseProps } from '../../interfaces/base';
import { IDropdown } from '../../interfaces/dropdown';
import { IAddress } from '../../interfaces/address';

interface IEmploymentStatuses extends IDropdown {}

export interface IEmploymentHistoryObjectSlim {
  employmentStatus: string;
  month: number | null;
  year: number | null;
}

export interface IEmploymentHistoryObjectFull
  extends IEmploymentHistoryObjectSlim {
  contractType: string;
  jobTitle: string;
  companyName: string;
  phoneNumber: string;
  address: IAddress | null;
  income: string;
}

export type TEmploymentHistory =
  | IEmploymentHistoryObjectSlim
  | IEmploymentHistoryObjectFull;

export interface IEmploymentHistoryProps extends IBaseProps {
  minYearsOfHistory: number;
  lookBackYears: number;
  employmentStatuses: IEmploymentStatuses;
  employmentHistory?: TEmploymentHistory[];
  setValue?: Function; // Callback.
}

export interface IEmploymentHistoryState {
  employmentHistory: TEmploymentHistory[];
  completed: boolean;
  allValid: boolean;
}
