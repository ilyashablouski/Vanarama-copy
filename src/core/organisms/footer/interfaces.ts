import { IBaseProps } from '../../interfaces/base';
import { ILink } from '../../interfaces/link';

export interface IFooterProps extends IBaseProps {
  emailAddress?: string;
  phoneNumber?: string;
  openingTimes?: IOpeningTimes[];
  headingZero?: string;
  headingOne?: string;
  headingTwo?: string;
  headingThree?: string;
  linkGroupOne?: ILink[];
  linkGroupTwo?: ILink[];
  linkGroupThree?: ILink[];
  legalText?: string;
}

export interface IOpeningTimes {
  day: string;
  time: string;
}
