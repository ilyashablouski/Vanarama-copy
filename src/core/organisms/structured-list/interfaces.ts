import { ChangeEvent } from 'react';
import { IBaseProps } from 'core/interfaces/base';
import { TSize } from '../../../types/size';
import { TColor } from '../../../types/color';

export interface IList extends IBaseProps {
  selectEdit?: boolean;
  textEdit?: boolean;
  placeholder?: string;
  dataTestId?: string;
  options?: {
    data: string[];
    favourites: string[];
  };
  name?: string;
  label: string;
  value: string | (string | JSX.Element)[];
  isOrange?: boolean;
  dataAbTestId?: string;
  labelElementAttributes?: IElementAttributes;
  valueElementAttributes?: IElementAttributes;
}

export interface IStructuredListProps extends IBaseProps {
  itemWrap?: boolean;
  list: IListItemProps[];
  editable?: boolean;
  heading?: string;
  headingSize?: TSize;
  priceTag?: { price: number; info: string; size?: TSize; color?: TColor };
  onChange?(event: ChangeEvent<HTMLSelectElement | HTMLInputElement>): void;
  /**
   * Callback function called when the user clicks the "Edit" link
   */
  onEditClicked?: () => void;
  /**
   * A data-testid attribute to attach to the "Edit" link
   */
  editDataTestId?: string;
  priceDataTestId?: string;
  priceTagDataTestId?: string;
  headingDataTestId?: string;
}

interface IElementAttributes {
  dataAbTestId: string;
}

export interface IListItemProps extends IBaseProps {
  id?: string;
  key?: string;
  selectEdit?: boolean;
  textEdit?: boolean;
  label: string;
  value: string | (string | JSX.Element)[];
  name?: string;
  placeholder?: string;
  options?: {
    data: string[];
    favourites: string[];
  };
  isOrange?: boolean;
  wrap?: boolean;
  editing?: boolean;
  testId?: number;
  dataAbTestId?: string;
  onChange?(event: ChangeEvent<HTMLSelectElement | HTMLInputElement>): void;
  labelElementAttributes?: IElementAttributes;
  valueElementAttributes?: IElementAttributes;
}
