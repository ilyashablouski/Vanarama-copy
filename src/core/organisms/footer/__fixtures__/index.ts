export const emailAddress = 'enquiries@vanarama.co.uk';

export const phoneNumber = '01442 838 195';

export const openingTimes = [
  {
    day: 'Monday–Thursday',
    time: '8:30am–7pm',
  },
  {
    day: 'Friday',
    time: '9am–5pm',
  },
  {
    day: 'Saturday',
    time: '9am–3pm',
  },
  {
    day: 'Sunday',
    time: 'Closed',
  },
];

export const legalText = `
Autorama UK Ltd Registered in England and Wales with registered
number: 05137709 Registered office: Vanarama, Maylands Avenue,
Hemel Hempstead, HP2 7DE, United Kingdom. Vanarama, Vanarama Cars,
Pickup Trucks Direct, Vanarama Short Term Leasing and Vanarama
Insurance Services are trading styles of Autorama UK Ltd, who are
authorised and regulated by the Financial Conduct Authority (Our
Financial Conduct Authority Register number is 630748). VAT number
GB842814720. Autorama UK Ltd is a Credit Broker not a Lender, we
can introduce you to a selected group of Lenders. Calls are
recorded for compliance and training purposes.`;

export const headingZero = 'Get In Touch With Us';

export const headingOne = 'Vans & Trucks';
export const linkGroupOne = [
  { label: 'Van Leasing', href: '#' },
  { label: 'Pickup Truck Leasing', href: '#' },
  { label: 'Van Reviews', href: '#' },
  { label: 'Van News', href: '#' },
  { label: 'Compare Vans', href: '#' },
  { label: 'Van Leasing Explained', href: '#' },
  { label: 'Ask Us A Question', href: '#' },
  { label: 'Advanced Breakdown Cover', href: '#' },
  { label: 'Service Plus', href: '#' },
  { label: 'Van Insurance', href: '#' },
  { label: 'Van Insurance FAQs', href: '#' },
];

export const headingTwo = 'Cars';
export const linkGroupTwo = [
  { label: 'Car Leasing', href: '#' },
  { label: 'Car News', href: '#' },
  { label: 'Car Leasing Explained', href: '#' },
  { label: 'Mileage Booster & Damage Cover', href: '#' },
  { label: 'Eligibility Checker', href: '#' },
  { label: 'Company Cash Allowance', href: '#' },
];

export const headingThree = 'About Us & Contact';
export const linkGroupThree = [
  { label: 'Our Price Promise', href: '#' },
  { label: '5 Star Service Promise', href: '#' },
  { label: 'Customer Leasing Reviews', href: '#' },
  { label: 'Award Winning Leasing', href: '#' },
  { label: 'Careers', href: '#' },
  { label: 'Our Local Representatives', href: '#' },
  { label: 'Community', href: '#' },
  { label: 'FCA Initial Disclosure Document', href: '#' },
  { label: 'Complaints Policy', href: '#' },
  { label: 'Cookie & Privacy Policy', href: '#' },
  { label: 'Terms & Conditions', href: '#' },
  { label: 'Modern Slavery Statement', href: '#' },
  { label: 'Sitemap', href: '#' },
];
