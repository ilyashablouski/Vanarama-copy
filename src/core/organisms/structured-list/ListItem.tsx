import React, { FC, ChangeEvent, useState } from 'react';

import Select from '../../atoms/select';
import TextInput from '../../atoms/textinput';

import { IListItemProps } from './interfaces';

const ListItem: FC<IListItemProps> = props => {
  const {
    testId,
    selectEdit,
    textEdit,
    label,
    value,
    name = '',
    placeholder,
    options,
    editing,
    onChange,
    dataTestId,
  } = props;

  const [input, setInput] = useState(value);

  const handleChange = (
    e: ChangeEvent<HTMLSelectElement | HTMLInputElement>,
  ) => {
    setInput(e.target.value);
    if (onChange) onChange(e);
  };

  const renderDefault = () => (
    <div className="structured-list-td" data-testid={dataTestId}>
      {typeof input === 'string'
        ? input
        : input.map(item => (
            <>
              {item}
              <br />
            </>
          ))}
    </div>
  );

  const renderTextInput = () => (
    <TextInput
      dataTestId={dataTestId}
      value={input}
      name={name}
      placeholder={placeholder}
      onChange={handleChange}
    />
  );

  const renderSelect = () => (
    <Select
      dataTestId={dataTestId}
      name={name}
      value={input as string}
      onChange={handleChange}
    >
      {options?.favourites.map(item => (
        <option key={item} value={item}>
          {item}
        </option>
      ))}
      {options?.data.map(item => (
        <option key={item} value={item}>
          {item}
        </option>
      ))}
    </Select>
  );

  const renderItem = () => {
    if (editing) {
      if (selectEdit) {
        return renderSelect();
      }

      if (textEdit) {
        return renderTextInput();
      }
    }

    return renderDefault();
  };

  return (
    <div className="structured-list-row ">
      <div
        className="structured-list-td structured-list-content--nowrap"
        data-testid={`data_id-${testId}-${label}`}
      >
        {label}
      </div>
      {renderItem()}
    </div>
  );
};

export default ListItem;
