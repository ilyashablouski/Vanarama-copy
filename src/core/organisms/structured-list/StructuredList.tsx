import React, { FC, useState } from 'react';
import cx from 'classnames';

import Price from '../../atoms/price';
import Text from '../../atoms/text';
import Link from '../../atoms/link';
import ListItem from './ListItem';

import { IStructuredListProps } from './interfaces';

const StructuredList: FC<IStructuredListProps> = ({
  className,
  list = [],
  editable = false,
  heading = '',
  onChange,
  headingSize,
  onEditClicked,
  priceTag,
  editDataTestId,
  priceDataTestId,
  priceTagDataTestId,
  headingDataTestId,
}) => {
  const [editing, setEditing] = useState(false);
  return (
    <section className={cx('structured-list', className)}>
      {(heading || priceTag) && (
        <div className="structured-list-thead">
          <div className="structured-list-row structured-list-row--header-row">
            {priceTag && (
              <div className="structured-list-th">
                <Price
                  price={priceTag.price}
                  size={priceTag.size}
                  dataTestId={priceDataTestId}
                />
                <Text
                  tag="p"
                  color={priceTag.color ? priceTag.color : 'black'}
                  size="xsmall"
                  dataTestId={priceTagDataTestId}
                >
                  {priceTag.info}
                </Text>
              </div>
            )}
            {heading && (
              <div className="structured-list-th">
                <Text
                  tag="p"
                  color="darker"
                  size={headingSize}
                  dataTestId={headingDataTestId}
                >
                  {heading}
                </Text>
              </div>
            )}
          </div>
        </div>
      )}
      <div className="structured-list-tbody">
        {list.map((item, index) => (
          <ListItem
            key={item.dataTestId}
            testId={index}
            onChange={onChange}
            editing={editing}
            dataTestId={item.dataTestId}
            {...item}
          />
        ))}
      </div>
      {editable && (
        <div className="structured-list--edit">
          {!editing ? (
            <Link
              color="teal"
              dataTestId={editDataTestId}
              onClick={() => {
                onEditClicked?.();
                setEditing(true);
              }}
            >
              Edit
            </Link>
          ) : (
            <Link color="teal" onClick={() => setEditing(false)}>
              Save Changes
            </Link>
          )}
        </div>
      )}
    </section>
  );
};

export default StructuredList;
