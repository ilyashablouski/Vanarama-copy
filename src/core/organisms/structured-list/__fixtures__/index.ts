import { IListItemProps } from 'core/organisms/structured-list/interfaces';

export const defaultList: IListItemProps[] = [
  {
    label: 'Email',
    value: 'example@example.com',
    id: 'email',
    dataTestId: 'email-value',
  },
  {
    label: 'Fullname',
    value: 'Mr John Doe',
    id: 'fullname',
    dataTestId: 'full-name-value',
  },
  {
    label: 'Mobile',
    value: '012100000',
    id: 'mobile',
    dataTestId: 'mobile-value',
  },
];

export const editableList: IListItemProps[] = [
  {
    textEdit: true,
    label: 'Email',
    value: 'example@example.com',
    name: 'email',
    id: 'email',
  },
  {
    textEdit: true,
    label: 'Fullname',
    value: 'Mr John Doe',
    name: 'fullname',
    id: 'fullname',
  },
  {
    textEdit: true,
    label: 'Mobile',
    value: '012100000',
    name: 'mobile',
    id: 'mobile',
  },
  {
    selectEdit: true,
    label: 'Country Of Birth',
    value: 'United Kingdom',
    options: {
      data: ['Unitied Kingdom', 'Canada', 'Unitied States'],
      favourites: ['United Kingdom'],
    },
    name: 'cob',
    id: 'cob',
  },
];
