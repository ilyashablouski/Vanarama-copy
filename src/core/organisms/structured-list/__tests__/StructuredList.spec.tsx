import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import StructuredList from '../StructuredList';
import ListItem from '../ListItem';
import { IStructuredListProps } from '../interfaces';
import { editableList, defaultList as list } from '../__fixtures__';

const getComponent = (props: IStructuredListProps) => (
  <StructuredList {...props} />
);

describe('<StructuredList />', () => {
  it('should render with required props', () => {
    const { container } = render(getComponent({ list }));

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a custom CSS class', () => {
    const { container } = render(
      getComponent({ list, className: 'custom-class' }),
    );

    expect(container).toMatchSnapshot();
  });

  it('should contain `<Link/>` with expected props/vals', () => {
    const wrap = render(
      getComponent({ list: editableList, editable: true }),
    ).container.getElementsByClassName('structured-list--edit')[0];

    expect(wrap.getElementsByClassName('link').length).toEqual(1);
    expect(wrap.getElementsByClassName('link')[0].textContent).toEqual('Edit');
    expect(
      fireEvent.click(wrap.getElementsByClassName('link')[0]),
    ).toBeTruthy();
  });

  it('should contain 4 ListItem components (defined in fixtures)', () => {
    const wrap = render(getComponent({ list: editableList, editable: true }))
      .container;

    expect(
      wrap
        .getElementsByClassName('structured-list-tbody')[0]
        .getElementsByClassName('structured-list-row').length,
    ).toEqual(4);
  });

  it('should contain `<ListItem/>` with default value (non editing state)', () => {
    const wrap = render(
      <ListItem textEdit editing={false} label="Test" value="test" />,
    ).container;
    expect(
      wrap.getElementsByClassName('structured-list-td')[1].textContent,
    ).toEqual('test');
  });

  it('should contain `<ListItem/>` with TextInput (editing state)', () => {
    const wrap = render(<ListItem textEdit editing label="Test" value="test" />)
      .container;
    expect(wrap.getElementsByClassName('textinput--native').length).toEqual(1);
  });

  it('should contain `<ListItem/>` with select (editing state)', () => {
    const wrap = render(
      <ListItem selectEdit editing label="Test" value="test" />,
    ).container;
    expect(wrap.getElementsByTagName('select').length).toEqual(1);
  });
  // TODO: fix test
  it.skip('should call `onEditClicked` when clicking the Edit link', () => {
    // // ARRANGE
    // const onEditClicked = jest.fn();
    //
    // // ACT
    // const wrapper = render(
    //   <StructuredList list={[]} editable onEditClicked={onEditClicked} />,
    // );
    //
    // // Find the anchor element with the text "Edit", and click it
    // wrapper.find('a[children="Edit"]').simulate('click');
    //
    // // ASSERT
    // expect(onEditClicked).toHaveBeenCalledTimes(1);
  });
  // TODO: fix test
  it.skip('should attach a data-testid attribute to the Edit link', () => {
    // ARRANGE
    // const dataTestId = 'my-data-testid';
    //
    // // ACT
    // const wrapper = mount(
    //   <StructuredList list={[]} editable editDataTestId={dataTestId} />,
    // );
    //
    // // ASSERT
    // // Find the anchor element with the text "Edit" and make sure it has a data-testid
    // expect(wrapper.find('a[children="Edit"]').prop('data-testid')).toEqual(
    //   dataTestId,
    // );
  });

  it('should attach a data-testid attributes the the list items', () => {
    // ARRANGE
    const dataTestId = 'my-data-testid';

    // ACT
    render(
      <StructuredList
        list={[
          { label: 'Some item label', value: 'Some item value', dataTestId },
        ]}
      />,
    );

    // ASSERT
    // Find the anchor element with the text "Edit" and make sure it has a data-testid
    expect(screen.getByTestId('my-data-testid').textContent).toEqual(
      'Some item value',
    );
  });
});
