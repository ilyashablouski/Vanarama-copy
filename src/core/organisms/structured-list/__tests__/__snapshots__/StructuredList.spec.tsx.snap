// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StructuredList /> should render correctly with a custom CSS class 1`] = `
<section
  className="structured-list custom-class"
>
  <div
    className="structured-list-tbody"
  >
    <ListItem
      dataTestId="email-value"
      editing={false}
      id="email"
      key="email-value"
      label="Email"
      testId={0}
      value="example@example.com"
    />
    <ListItem
      dataTestId="full-name-value"
      editing={false}
      id="fullname"
      key="full-name-value"
      label="Fullname"
      testId={1}
      value="Mr John Doe"
    />
    <ListItem
      dataTestId="mobile-value"
      editing={false}
      id="mobile"
      key="mobile-value"
      label="Mobile"
      testId={2}
      value="012100000"
    />
  </div>
</section>
`;

exports[`<StructuredList /> should render with required props 1`] = `
<section
  className="structured-list"
>
  <div
    className="structured-list-tbody"
  >
    <ListItem
      dataTestId="email-value"
      editing={false}
      id="email"
      key="email-value"
      label="Email"
      testId={0}
      value="example@example.com"
    />
    <ListItem
      dataTestId="full-name-value"
      editing={false}
      id="fullname"
      key="full-name-value"
      label="Fullname"
      testId={1}
      value="Mr John Doe"
    />
    <ListItem
      dataTestId="mobile-value"
      editing={false}
      id="mobile"
      key="mobile-value"
      label="Mobile"
      testId={2}
      value="012100000"
    />
  </div>
</section>
`;
